designs/node-templates-and-drift.md

```diff
@@ -0,0 +1,109 @@
+# Karpenter Deprovisioning - Node Template and Drift
+
+Drift is implemented to assert that nodes are running with the AMIs as specified in their AWSNodeTemplate. Karpenter has many more configuration options that must be considered for Drift. These are segmented by Provisioner Drift and Provider Drift. This document is to validate that drift on each configuration option for the Provisioner and Provider make sense, and how to reason about future API fields.
+
+## Provisioners and Deployments
+
+Analogous to the [relationship of Deployment -> Pods](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template), nodes are owned by the Machine which is owned by the Provisioner, nodes have the provisioner-name label in the metadata, and nodes are heterogenous replicas of the Provisioner’s requirements. A node refers to only one Provisioner, and each Provisioner refers to only one AWSNodeTemplate. In this, we can create logical mapping as follows.
+
+![](images/provisioner-node-relationship.png)
+
+Drift is actually already implemented in the Kubernetes Deployment object for pods. Given the label selector of Pods in a [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector), Kubernetes “may terminate Pods whose labels match the selector if their template is different from `.spec.template`”. More specifically, Kubernetes maintains a hash of the pod template, annotates the pods with that hash, and then uses it as a checksum to ensure they remain the same. Finally, Kubernetes even allows [strategy](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy) for how to roll out deployment changes when pods drift from its deployment.
+
+## 🔑 ProvisionerSpec fields → ProvisionerSpec.template
+
+```
+spec:
+  template:
+    requirements: {...}
+    labels: {...}
+    taints: {...}
+    annotations: {...}
+    startupTaints: {...}
+    providerRef: {...}
+```
+
+Following this semantic, the `ProviderRef`/`in-line Provider` struct will be included in the template. The `ProviderRef`/`Provider` is responsible for any Cloud Provider-specific Machine Template requirements and will also be nested within template.
+
+## How does this relate to Drift?
+
+A machine will be drifted if the machine’s in-memory template matches the by the `ProvisionerSpec.Template` and `Provider.Spec` fields.
```

Comment:
I suspect `the by the` is a typo.

But the sentence as a whole doesn't make sense to me - I expect things to drift when things don't match.

Reply:
Good catches, will fix. Thanks!